### PR TITLE
#138 add support for collapse

### PR DESCRIPTION
**Pull request by vfrank66**

### CODE REVIEW

File: tests/enginetests/operator_interval_test.go [hunk 1]
```diff
@@ -125,6 +125,89 @@ func TestEnd(t *testing.T) {
...
+						LowInclusive:  true,
+						HighInclusive: true,
+						StaticType:    &types.Interval{PointType: types.Date},
+					}),
+				},
+				StaticType: &types.List{ElementType: &types.Interval{PointType: types.Date}},
+			}),
+		},
+		{
+			name: "Collapse empty list",
```
Comment (evan-gordon) [inline tests/enginetests/operator_interval_test.go:167]: This test name is wrong, technically this is null, an empty list would be {} as ...

We would probably want tests for both though :)

File: tests/enginetests/operator_interval_test.go [hunk 2]
```diff
@@ -125,6 +125,89 @@ func TestEnd(t *testing.T) {
...
+					newOrFatal(t, result.Interval{
+						Low:           newOrFatal(t, int32(1)),
+						High:          newOrFatal(t, int32(8)),
+						LowInclusive:  true,
+						HighInclusive: true,
+						StaticType:    &types.Interval{PointType: types.Integer},
+					}),
+				},
+				StaticType: &types.List{ElementType: &types.Interval{PointType: types.Integer}},
+			}),
```
Comment (evan-gordon) [inline tests/enginetests/operator_interval_test.go:188]: Could you add some tests for cases that shouldn't be collapsed, both with closed and open intervals?

File: tests/enginetests/operator_interval_test.go [hunk 3]
```diff
@@ -1966,3 +2049,308 @@ func TestComparison_Error(t *testing.T) {
...
+		},
+		{
+			name: "Original failing case - sorted order",
```
Comment (evan-gordon) [inline tests/enginetests/operator_interval_test.go:2316]: I believe you already have this case on line 2164 consider removing.

File: tests/spectests/exclusions/exclusions.go [hunk 4]
```diff
@@ -240,6 +239,9 @@ func XMLTestFileExclusionDefinitions() map[string]XMLTestFileExclusions {
 				"Union",
 			},
 			NamesExcludes: []string{
+				"TestCollapseNull",
```
Comment (evan-gordon) [inline tests/spectests/exclusions/exclusions.go:242]: just curious, what error were you getting for this test?

File: interpreter/operator_interval.go [hunk 5]
```diff
@@ -641,3 +642,1183 @@ func evalWidthInterval(m model.IUnaryExpression, intervalObj result.Value) (resu
...
+
+	// Remove null intervals first
+	var nonNullIntervals []result.Value
+	for _, interval := range intervals {
+		if !result.IsNull(interval) {
+			nonNullIntervals = append(nonNullIntervals, interval)
+		}
+	}
+
+	if len(nonNullIntervals) == 0 {
```
Comment (evan-gordon) [inline interpreter/operator_interval.go:1077]: You could also return early here if the length is 1.

File: interpreter/operator_interval.go [hunk 6]
```diff
@@ -641,3 +642,1183 @@ func evalWidthInterval(m model.IUnaryExpression, intervalObj result.Value) (resu
...
+		return result.New([]result.Value{})
+	}
+
+	// Sort temporal intervals by start time
+	sort.Slice(nonNullIntervals, func(i, j int) bool {
```
Comment (evan-gordon) [inline interpreter/operator_interval.go:1082]: Can't you use the helper function you added above for this?

File: interpreter/operator_interval.go [hunk 7]
```diff
@@ -641,3 +642,1183 @@ func evalWidthInterval(m model.IUnaryExpression, intervalObj result.Value) (resu
...
+		return result.Value{}, err
+	}
+	collapsedIntervals = append(collapsedIntervals, cleanLastInterval)
+
+	// Return the collapsed intervals with the correct list type
+	return result.New(result.List{Value: collapsedIntervals, StaticType: listType})
+}
+
+// sortIntervals sorts a list of intervals by their start time
+func sortIntervals(intervals []result.Value, evaluationTimestamp time.Time) ([]result.Value, error) {
```
Comment (evan-gordon) [inline interpreter/operator_interval.go:1006]: I like where this function is going but I have a few thoughts.
1) I think sortIntervalsByStartTime would be a more apt name here.
2) What do you think of using generics here similar to https://github.com/google/cql/blob/25fd8b5c1cb27c77f86e3e13aeab99560a9b4411/interpreter/operator_arithmetic.go#L321
since sorting for time based values and numerals are such different operations it may make sense to split the two implementations out in this way.

File: interpreter/operator_interval.go [hunk 8]
```diff
@@ -641,3 +642,1183 @@ func evalWidthInterval(m model.IUnaryExpression, intervalObj result.Value) (resu
...
+			nonNullIntervals = append(nonNullIntervals, interval)
+		}
+	}
+
+	if len(nonNullIntervals) == 0 {
+		return result.New([]result.Value{})
+	}
+
+	// Sort numeric intervals by start value
+	sort.Slice(nonNullIntervals, func(i, j int) bool {
```
Comment (evan-gordon) [inline interpreter/operator_interval.go:1418]: same feedback as the above function i think you can reuse your helper here

File: tests/enginetests/operator_interval_test.go [hunk 9]
```diff
@@ -125,6 +125,89 @@ func TestEnd(t *testing.T) {
 	}
 }
 
+func TestCollapse(t *testing.T) {
```
Comment (evan-gordon) [inline tests/enginetests/operator_interval_test.go:128]: I think some of the logic for `null` values like `Interval[null, 5]` or `Interval(null, 5]` might not be handled correctly, at least the canMerge functions seem to be marking them all as false for their ability to merge. Mind adding some test cases like`collapse { Inverval[null, 5], Interval[1, 2] }` as well as `collapse { Inverval(null, 5], Interval[1, 2] }` and `collapse { Inverval[null, 5], Interval[4, 6] }`? I'd feel better about this code knowing those do what I expect :)

**Comment by evan-gordon** [comment]
Since there's been no activity here for a while but this change is mostly there I'm going to conditionally approve it and then manually address the current comments.
